components/sup/src/manager/service/hooks.rs

```diff
@@ -20,6 +19,25 @@ use std::{self,
 
 static LOGKEY: &str = "HK";
 
+#[derive(Debug)]
+pub struct CompleteHookOutput {
+    stdout:      Option<String>,
+    stderr:      Option<String>,
+    exit_status: ExitStatus,
+}
+
+impl CompleteHookOutput {
+    fn new(hook_output: &HookOutput, exit_status: ExitStatus) -> Self {
+        Self { stdout: hook_output.stdout_str(),
+               stderr: hook_output.stderr_str(),
+               exit_status }
+    }
+
+    pub fn get_exit_status(&self) -> &ExitStatus { &self.exit_status }
+
+    pub fn output_streams(self) -> (Option<String>, Option<String>) { (self.stdout, self.stderr) }
```

Comment:
This is a bit of a misnomer since the returned types aren't streams, but strings. And since a tuple of identical types is returned, it's easy for the caller to make a mistake by reversing them. I see two ways to make this safer:

1. Have separate methods for each like:
    ```rust
    pub fn stdout(&mut self) -> Option<String> {
        self.stdout.take()
    }
    ```
    Then the caller becomes:
    ```rust
    let (stdout, stderr) = hook_output
        .map(|mut o| (o.stdout(), o.stderr()))
        .unwrap_or_default();
     ```
1. Create a type which can be destructured like:
    ```rust
    #[derive(Default)]
    pub struct HookTextOutput {
        pub stdout: Option<String>,
        pub stderr: Option<String>,
    }
    ```
    Then the caller becomes:
    ```rust
    let HookTextOutput { stdout, stderr } = hook_output
        .map(CompleteHookOutput::text_output)
        .unwrap_or_default();
    ```

I think the latter is better since it's even a bit safer and fairly concise. We could even define `CompleteHookOutput` in terms of `HookTextOutput` without losing encapsulation:
```rust
#[derive(Debug)]
pub struct CompleteHookOutput {
    out: HookTextOutput,
    exit_status: ExitStatus,
}
```
since even though the fields of `HookTextOutput` are `pub`, the `out` field of `CompleteHookOutput` is not.

Reply:
"Stream" is an overloaded term. I was meaning in the linux [standard stream](https://en.wikipedia.org/wiki/Standard_streams) sense. Not sure of a better term. Maybe `process_streams`?

I agree with the change to make the return value safer.

Reply:
`process_outputs`?

Reply:
With this [commit](https://github.com/habitat-sh/habitat/pull/6905/commits/ea08605617b9e5ac58936da60cc0b7d2d32ea30f), I switched `CompleteHookOutput` to `ProcessOutput` and used `standard_stream` instead of `output_stream`. If there is concensus that we prefer `process_output` over `standard_stream` I will switch to that.